crates/driver/src/domain/competition/solution/mod.rs

```diff
@@ -284,49 +289,77 @@ impl SolverTimeout {
     }
 }
 
-/// The solution score. This is often referred to as the "objective value".
-#[derive(Debug, Default, PartialEq, Eq, PartialOrd, Ord, Copy, Clone)]
-pub struct Score(pub eth::U256);
+/// Carries information how the score should be calculated.
+#[derive(Debug, Clone)]
+pub enum Score {
+    Solver(eth::U256),
+    Discount(eth::U256),
+    RiskAdjusted {
+        success_probability: SuccessProbability,
+        gas_amount: Option<eth::U256>,
+    },
+}
 
-impl From<Score> for eth::U256 {
-    fn from(value: Score) -> Self {
-        value.0
-    }
+/// Represents the probability that a solution will be successfully settled.
+#[derive(Debug, Clone)]
+pub enum SuccessProbability {
+    /// Probability known and is equal to the given value.
+    Value(f64),
+    /// Probability unknown and should be computed by the protocol using the
+    /// given parameters.
+    Params {
+        gas_amount_factor: f64,
+        gas_price_factor: f64,
+        nmb_orders_factor: f64,
+        intercept: f64,
+    },
```

Comment:
I mentioned this [here](https://cowservices.slack.com/archives/C0375NV72SC/p1695306427969929)
But we don't have the simulation implemented in the `solvers` crate right now.

Reply:
If I understand correctly we need the gas amount of the solution (not necessarily a full simulation). If I understand correctly we already have this information somehow (e.g. to compute the fee on partially fillable limit orders). While maybe not 100% accurate I think we should use whatever we use for fee computation here as well.

Reply:
> If I understand correctly we need the gas amount of the solution

Yes.

> Don't we already have this information somehow (e.g. to compute the fee on partially fillable limit orders)

I know at least Naive solver is not supporting limit orders: https://github.com/cowprotocol/services/blob/main/crates/solver/src/solver/naive_solver.rs#L67
We would either have to drop Naive or implement missing fee calculation. 
For `SingleOrderSolver`s we also do simulation to estimate `gas_amount`... maybe some of them provide gas via api response but probably not all...

Reply:
I might have given too much responsibility to the `solvers` crate in my head. After some more thinking, if we conclude that it will never be practical (to accurately) estimate gas_amount and do simulations in `solvers`, than RiskCalculator does actually belong to `driver`. Non-accurate gas estimation might come from pre/post hooks, access_lists, merging settlements logic, submission settlement configuration etc... and all of these are in the `driver` yard.
If so, than asking external solvers to provide accurate `success_probability` is under question as well(right?).

Right now I am totally undecided. One thing I would really hate is to have to extract the solver type from the api route so just I could know which solver provided the solution (or to send solver type within each solution from `solvers`). It seems to me that `driver` should not care about that, it should just care about the solution and all solution data that is specific to some solver should already be prepared by `solvers` for `driver` to consume.

But my thinking might be naive, because we would probably need solver type in `driver` for something sooner or later.
@fleupold @nlordell @MartinquaXD

Reply:
> One thing I would really hate is to have to extract the solver type from the api route so just I could know which solver provided the solution (or to send solver type within each solution from solvers). It seems to me that driver should not care about that, it should just care about the solution and all solution data that is specific to some solver should already be prepared by solvers for driver to consume.

The existing code in the driver is already aware about solver specific [configurations](https://github.com/cowprotocol/services/blob/main/crates/driver/src/infra/solver/mod.rs#L61-L70) and they are easily available in the route handlers of the driver. I think the most convincing argument why this makes sense is that the `solver` itself doesn't care about how the solution gets on-chain. Therefor it would be kind of weird to make it responsible for telling the `driver` its private key so that it can submit the solution. That's 100% the `driver`'s responsibility so why not have the private key in a per-solver configuration?

I think the best way to think about the `solver` is that it's really just crunching some numbers to produce a solution in **some** format and the `driver` is supposed to make it so that the protocol can actually understand the format and submit the solution. With that way of thinking we should be able to compute the success probability because after all 2 completely different solving algorithms (i.e. solvers) might produce the exact same solution. The exact same solution should have the exact same success probability which means it does not inherently depend on anything solver specific.

The only thing I'm currently worried about is that the score computation so far seems to be the only piece of logic in the driver which I can reasonably see people wanting to adjust to their liking (I might be wrong on this, though). If that logic lives in the driver it will be harder for people to adjust than if they have full control in the solver. However, I already argued that the proper place for that logic seems to be in the driver. So how do we reconcile these things?
1. have a single built-in function for the success probability
2. offer multiple different ones
3. somehow allow people to define their own logic
4. allow returning `Option<SuccessProbability>` with the solution (and return `None` for all the gnosis implementations to make the driver compute it)
5. something else entirely

Currently I'm leaning towards `4` because it's just a tiny "wart" in the API which buys **a lot** of flexibility. Technically we could also allow setting a `score` but that seems a bit silly because there is an optimal way to compute the score an solvers seemed to struggle with that already.

Reply:
Calculating `success_probability` is moved to `solvers`.
Currently the `success_probability: 1.0` is returned to `driver` for naive, baseline, dex.
Proper calculation for those is part of the https://github.com/cowprotocol/services/issues/1893